vhost-device-console: add Unix domain socket backend support (#821)
rust-vmm/vhost-device: vhost-device-console/README.md

 The localhost's port to be used for each guest, this part will be increased with
 0,1,2..socket_count-1.
 And the option is only valid when backend mode is "network".

-- option:: -b, --backend=nested|network|uds

  The backend type vhost-device-console to be used. The current implementation
  supports 3 types of backends: "nested", "network", "uds".
  Note: The nested backend is selected by default and can be used only when
        socket_count equals 1.

.. option:: --uds-path=uds-file-path

  The unix domain socket to be used for each guest, this path will be suffixed with
  0,1,2..socket_count-1. e.g.: `--uds-path=/tmp/vm.sock --socket-count=2`
  leads to two connectable vhost-user sockets:
  /tmp/vm.sock0, /tmp/vm.sock1

>   And the option is only valid when backend mode is "uds".

Reviewer: ```suggestion
  This option is only valid when backend type is "uds".
```

Author: done

 The localhost's port to be used for each guest, this part will be increased with
 0,1,2..socket_count-1.

-- option:: -b, --backend=nested|network|uds

  The backend type vhost-device-console to be used. The current implementation
  supports 3 types of backends: "nested", "network", "uds".
  Note: The nested backend is selected by default and can be used only when
        socket_count equals 1.

> .. option:: -v, --uds-path=uds-file-path

Reviewer: Why the short is `-v` ?
I'm not a fan of short option, but if really want one, I'd say `-u` makes more sense.

Author: delete the short option.

  The unix domain socket to be used for each guest, this path will be suffixed with
  0,1,2..socket_count-1. e.g.: `--uds-path=/tmp/vm.sock --socket-count=2`
  leads to two connectable vhost-user sockets:

>   /tmp/vm.sock0, /tmp/vm.sock1

Reviewer: Please make it clear that this parameter makes sense on with `uds` backend

Author: ok

The daemon should be started first:
```shell
host# vhost-device-console --socket-path=/tmp/console.sock --socket-count=1 \
                           --tcp-port=12345 --backend=network # for network backend
or

host# vhost-device-console --socket-path=/tmp/console.sock --socket-count=1 \
                           --uds-path=/tmp/vm.sock --backend=uds # for uds backend
```
>Note: In case the backend is "nested" there is no need to provide
       "--socket-count", "--tcp-port" and "--uds-path" parameters.

The QEMU invocation needs to create a chardev socket the device can
use to communicate as well as share the guests memory over a memfd.

> There are two option for running QEMU with vhost-device-console,

Reviewer: Here we are talking about QEMU, please add a new section for UML (and maybe also one for QEMU with the other info we already have), it's a bit confusing.

Author: Add a new configuration section for UML. No changes are needed for QEMU, as QEMU does not recognize this parameter.

    ...
```

3) Using User Mode Linux [`virtio console`](https://github.com/torvalds/linux/blob/848e076317446f9c663771ddec142d7c2eb4cb43/include/uapi/linux/virtio_ids.h#L34):
```text
host# linux root=/dev/ubda1 rw ubd0=$YOUR-PATH/kata-ubuntu-latest.image   \
    <normal UML options>                                             \
    virtio_uml.device=/tmp/console.sock0:3 console=tty0 console=hvc0 \
    init=/bin/systemd                                                \
    systemd.unit=kata-containers.target agent.debug_console
```
Test with [kata-ubuntu-latest.image](https://github.com/kata-containers/kata-containers/releases/),
you can also use systemd to setup pty manually without kata-agent help.

Eventually, the user can connect to the console by running:
```test
host# stty -icanon -echo && nc localhost 12345 && stty echo # for network backend

> or

Reviewer: put the or outside the block or create 2 points

Author: ok

    ...
```

#### Test the device with UML
Start the daemon as above section, then run the following cmdline for
Kernel Mode Linux [`virtio console`](https://github.com/torvalds/linux/blob/848e076317446f9c663771ddec142d7c2eb4cb43/include/uapi/linux/virtio_ids.h#L34):
```text
host# linux root=/dev/ubda1 rw ubd0=$YOUR-PATH/kata-ubuntu-latest.image   \
    <normal UML options>                                             \
    virtio_uml.device=/tmp/console.sock0:3 console=tty0 console=hvc0 \
    init=/bin/systemd                                                \

>     systemd.unit=kata-containers.target agent.debug_console

Reviewer 2: I think testing with UML does not need to depend on Kata Containers:

- Remove staff regarding kata-containers: `systemd.unit`, image, `debug_console`.
- Use `/bin/bash` instead.

```shell
$ sudo linux ubd0=<rootfs> \
    init=/bin/bash \
    console=hvc0 \
    virtio_uml.device=/tmp/console.sock0:3
```

Reviewer 2: BTW, Kata's rootfs image can still be used here.
